hyperspy/tests/mva/test_bss.py

```diff
@@ -0,0 +1,197 @@
+        for c2 in c2_list:
+            if (np.allclose(c2.data, c1.data, atol=atol) or
+                    np.allclose(c2.data, -c1.data, atol=atol)):
+                matches += 1
+    return matches == len(c1_list)
+
+
+class TestBSS1D:
+
+    def setUp(self):
+        ics = np.random.laplace(size=(3, 1000))
+        np.random.seed(1)
+        mixing_matrix = np.random.random((100, 3))
+        self.s = Spectrum(np.dot(mixing_matrix, ics))
+        self.s.decomposition()
+
+    def test_on_loadings(self):
+        self.s.blind_source_separation(
+            3, diff_order=0, fun="exp", on_loadings=False)
+        s2 = self.s.as_spectrum(0)
+        s2.decomposition()
+        s2.blind_source_separation(
+            3, diff_order=0, fun="exp", on_loadings=True)
+        nose.tools.assert_true(are_bss_components_equivalent(
+            self.s.get_bss_factors(), s2.get_bss_loadings()))
+
+    def test_mask_diff_order_0(self):
+        mask = self.s._get_signal_signal(dtype="bool")
+        mask[5] = True
+        self.s.learning_results.factors[5, :] = np.nan
+        self.s.blind_source_separation(3, diff_order=0, mask=mask)
+
+    def test_mask_diff_order_1(self):
+        mask = self.s._get_signal_signal(dtype="bool")
+        mask[5] = True
+        self.s.learning_results.factors[5, :] = np.nan
+        self.s.blind_source_separation(3, diff_order=1, mask=mask)
+
+    def test_mask_diff_order_0_on_loadings(self):
+        mask = self.s._get_navigation_signal(dtype="bool")
+        mask[5] = True
+        self.s.learning_results.loadings[5, :] = np.nan
+        self.s.blind_source_separation(3, diff_order=0, mask=mask,
+                                       on_loadings=True)
+
+    def test_mask_diff_order_1_on_loadings(self):
+        mask = self.s._get_navigation_signal(dtype="bool")
+        mask[5] = True
+        self.s.learning_results.loadings[5, :] = np.nan
+        self.s.blind_source_separation(3, diff_order=1, mask=mask,
+                                       on_loadings=True)
+
+
+class TestBSS2D:
+
+    def setUp(self):
+        ics = np.random.laplace(size=(3, 1024))
+        np.random.seed(1)
+        mixing_matrix = np.random.random((100, 3))
+        s = Image(np.dot(mixing_matrix, ics).reshape((100, 32, 32)))
+        for (axis, name) in zip(s.axes_manager._axes, ("z", "y", "x")):
+            axis.name = name
+        s.decomposition()
+        self.s = s
+
+    def test_diff_axes_string_with_mask(self):
+        factors = self.s.get_decomposition_factors().inav[:3].diff(
+            axis="x", order=1)
+        self.s.blind_source_separation(
+            3, diff_order=0, fun="exp", on_loadings=False, factors=factors)
+        matrix = self.s.learning_results.unmixing_matrix.copy()
+        print matrix
+        self.s.blind_source_separation(
+            3, diff_order=1, fun="exp", on_loadings=False,
+            diff_axes=["x"],
+        )
+        matrix = self.s.learning_results.unmixing_matrix.copy()
```

Comment:
This line should be deleted, as it overwrites the results of previous calculations, hence in the end the test checks that  a copy is equal to the original object